public/app/features/dashboard/components/DashboardPrompt/DashboardPrompt.tsx

```diff
@@ -0,0 +1,214 @@
+            setState({ ...state, modal: null });
+          }}
+        />
+      )}
+      {modal === PromptModal.SaveLibraryPanelModal && (
+        <SaveLibraryPanelModal
+          panel={dashboard.panelInEdit as PanelModelWithLibraryPanel}
+          folderId={dashboard.meta.folderId as number}
+          onConfirm={() => {
+            locationService.push(blockedLocation!);
+          }}
+          onDiscard={() => {
+            dispatch(discardPanelChanges());
+            locationService.push(blockedLocation!);
+          }}
+          onDismiss={() => {
+            setState({ ...state, modal: null });
+          }}
+        />
+      )}
+    </>
+  );
+});
+
+/**
+ * For some dashboards and users changes should be ignored *
+ */
+function ignoreChanges(current: DashboardModel, original: object | null) {
+  if (!original) {
+    return true;
+  }
+
+  // Ignore changes if the user has been signed out
+  if (!contextSrv.isSignedIn) {
+    return true;
+  }
+
+  if (!current || !current.meta) {
+    return true;
+  }
+
+  const { canSave, fromScript, fromFile } = current.meta;
+  if (!contextSrv.isEditor && !canSave) {
+    return true;
+  }
+
+  return !canSave || fromScript || fromFile;
+}
+
+/**
+ * Remove stuff that should not count in diff
+ */
+function cleanDashboardFromIgnoredChanges(dashData: any) {
+  // need to new up the domain model class to get access to expand / collapse row logic
+  const model = new DashboardModel(dashData);
+
+  // Expand all rows before making comparison. This is required because row expand / collapse
+  // change order of panel array and panel positions.
+  model.expandRows();
+
+  const dash = model.getSaveModelClone();
+
+  // ignore time and refresh
+  dash.time = 0;
+  dash.refresh = 0;
+  dash.schemaVersion = 0;
+  dash.timezone = 0;
+
+  // ignore iteration property
+  delete dash.iteration;
+
+  dash.panels = filter(dash.panels, (panel) => {
+    if (panel.repeatPanelId) {
+      return false;
+    }
+
+    // remove scopedVars
+    panel.scopedVars = undefined;
+
+    // ignore panel legend sort
+    if (panel.legend) {
+      delete panel.legend.sort;
+      delete panel.legend.sortDesc;
+    }
+
+    return true;
+  });
+
+  // ignore template variable values
+  each(dash.getVariables(), (variable: any) => {
+    variable.current = null;
+    variable.options = null;
+    variable.filters = null;
+  });
+
+  return dash;
+}
+
+function hasChanges(current: DashboardModel, original: any) {
+  const currentClean = cleanDashboardFromIgnoredChanges(current.getSaveModelClone());
+  const originalClean = cleanDashboardFromIgnoredChanges(original);
+
+  const currentTimepicker: any = find((currentClean as any).nav, { type: 'timepicker' });
+  const originalTimepicker: any = find((originalClean as any).nav, { type: 'timepicker' });
+
+  if (currentTimepicker && originalTimepicker) {
+    currentTimepicker.now = originalTimepicker.now;
+  }
+
+  const currentJson = angular.toJson(currentClean);
```

Comment:
nit: can we use `safeStringifyValue` instead of angular.toJson?

Reply:
just copied from ChangeTracker